ui/src/components/Publication/Visualization/index.tsx

```diff
@@ -242,6 +207,137 @@ const Visualization: React.FC<VisualizationProps> = (props): React.ReactElement
+        }
+        Object.assign(publicationsByType, orderedParents);
+
+        const orderedChildren: {
+            [key in Types.PublicationType]?: Interfaces.LinkedFromPublication[];
+        } = {};
+        let childTypeIdx = filteredPublicationTypes.indexOf(selectedPublication.type) + 1;
+        // For each type, going forwards towards REAL_WORLD_APPLICATION, starting with the one after the selected publication's type...
+        while (childTypeIdx <= filteredPublicationTypes.indexOf('REAL_WORLD_APPLICATION')) {
+            const type = filteredPublicationTypes[childTypeIdx];
+            const publicationsOfType = linkedFrom.filter((linkedPublication) => linkedPublication.type === type);
+
+            // Sort child publications.
+            // For the type immediately after the selected publication's type, just order children by publication date, descending.
+            if (childTypeIdx === filteredPublicationTypes.indexOf(selectedPublication.type) + 1) {
+                orderedChildren[type] = publicationsOfType.sort(
+                    (a, b) => new Date(b.publishedDate).valueOf() - new Date(a.publishedDate).valueOf()
+                );
+            } else {
+                // For types further along the chain, order them to keep the visualisation as neat as we can.
+                // Pass the publications of the type we want to sort, and the (sorted) publications of the type before that.
+                const precedingType = filteredPublicationTypes[childTypeIdx - 1];
+                const precedingTypePublicationsOrdered = orderedChildren[precedingType];
+                if (precedingTypePublicationsOrdered !== undefined) {
+                    orderedChildren[type] = getOrderedLinkedPublications(
+                        publicationsOfType,
+                        precedingTypePublicationsOrdered
+                    ) as Interfaces.LinkedFromPublication[];
+                }
+            }
+            childTypeIdx++;
+        }
+        Object.assign(publicationsByType, orderedChildren);
+        // Set up the data we need to render a box for each publication.
+        for (const type of filteredPublicationTypes) {
+            const publicationsOfType = publicationsByType[type];
+            const boxEntriesOfType: BoxEntry[] = [];
+            if (publicationsOfType !== undefined) {
+                for (const publication of publicationsOfType) {
+                    // Get ids of children of this publication in the chain to which we will render a pointer arrow.
+                    const pointers =
+                        // This is the selected publication or its descendant, so get children from linkedFrom.
+                        type === selectedPublication.type ||
+                        filteredPublicationTypes.indexOf(type) >
+                            filteredPublicationTypes.indexOf(selectedPublication.type)
+                            ? [
+                                  ...new Set(
```

Comment:
Ordering the visualization boxes seems to add a fair amount of complexity with all these nested loops. This can have a negative impact on performance and ultimately affect users devices as the links count grows.

If possible (without too much hassle), I recommend having the links ordering logic (and any other hard work like this) on the backend and leave the UI as simple as possible so that it doesn't affect user's device performance.

Reply:
I've just pushed a commit which moves the bulk of this to the publication service in the API